optimizely/marketing-website
# Add webpack

This PR addresses adding webpack module loader into the marketing repo.  I know this implementation isn't a very "webpack" way, but I'm first trying to mimic our build behavior that was happening with the `concat` task to get everything working, and then go back and refactor. 

#### Issues:
- [x] Build does not use `grunt-webpack` but rather the webpack API within a `grunt.registerTask`. This was much more advantageous for usage of env configs appended on the grunt object that the webpack task depends upon.
- [x] Add inline sourcemap support. Must add the `website-guts` directory to the [chrome dev tools workspace](https://developer.chrome.com/devtools/docs/workspaces).
- [x] Dev and production config, in prod Uglify and Dedupe, in dev add the Hot Module Replacement plugin in preparation to convert to Webpack Dev Server for dev livereload.
- [x] Bundling vendor files (along with global JS) into one file and making their globals available outside of the bundle.  Currently, I'm accomplishing this through the `script-loader`, which seems to concat all files into a string and call `eval` on it.  Libraries are required directly from Bower in `global.js` and bundles with all other global files.
- [x] Mimicking `concat` behavior for the `namespacePages` target https://github.com/optimizely/marketing-website/blob/master/grunt/concat.js. i.e. separate bundles will be created for `page` and `layout` level JS and they will be output to `dist/assets/js` directory in the appropriate directory structure `layout/*.js` or `page/*.js`.  
- [x] Create banner and footer wrapper using `inject-filename-loader` which injects a `targetName` variable for JS error logging of bundles in try/catch blocks in staging/prod.
- [ ] Handlebars is being bundled more than once.  Once in global.js and additional times in page/layout js. Find a way to have handlebars only bundles once, or bundled globally and not included in the hbs-loader requires. Was recommended the [Common Chunks Plugin](https://github.com/webpack/docs/wiki/list-of-plugins#user-content-commonschunkplugin) ( not merge blocking )
- [x] Move JSHint into webpack. All JSHint is happening through webpack except for the linting the test and server side files. JSHint config for all tasks is still necessary as it is `require`d in the `webpack` grunt task.
- [x] Om (PPC) pages not using webpack, still using concat. This is because these must support very old browsers and must maintain their separate build environment.

### To Test

make sure to `rm -rf node_modules && npm cache clear && npm i`
`rm -rf bower_components && bower cache clean && bower i`
run `grunt server`
- watch the build, that webpack compiles correctly. It will be halted on any JSHint errors.
- inspect your console to make sure there are no JS errors
- inspect the dist directory and see that all of the JS files have an associate inline source map and that individual modules are wrapped in the closure transforming window and document to `w` and `d`, respectively.
- test a page like `/events` that uses a client side handlebars template. These are now being compiled with the webpack handlebars loader.
- in the `grunt/webpack/index.js` file find the `opts` object.  Append a `console.log(targetName)` to the end of the query string 
```js
injectFileNameParams: [
        'inject=' + encodeURIComponent('var targetName = __filename.replace("website-guts/assets/js/", "");\n\nconsole.log(targetName);\n\n'),
        '&banner=' + encodeURIComponent(banner),
        '&footer=' + encodeURIComponent(footer)
      ].join('')
```
run the `grunt server` task again.  You will see all individual JS module file names logged.  This is what is used in error reporting to GA and debugging in production.

- check [Staging](https://www.optimizelystaging.com/dfoxpowell/webpack-integration/pricing/)
- look in `website-guts/assets/js/global.js` and see how the `require` functions bootstrap the JS in a particular order and omits certain files using a RegEx with `require.context`.  This order is important for internal dependencies
- play around with some es6 if you want
- throw an error in the JS and check that it maps back to the appropriate source file through the source map http://cl.ly/27173s2T3p00
- run grunt build and inspect that inline source maps are not present and that individual modules are wrapped in try/catch blocks and that code has been uglified
- run the tests: 
```
grunt server
grunt ui-test //in a separate tab

grunt test //run this instead of the two commands above
```

### Enhancements
- we can now write es6
- source maps for local JS debugging
- JSHint inside webpack
- elimination of handlebars task - handlebars compiled on the fly in webpack loader. Also webpack loader handlebars has the same syntax as Assemble v6 for handlebars helpers which will be helpful for @4vanger `tr` helper.
- drastic reduction of code in concat, JSHint, and uglify task
- Handlebars is no longer required in the vendor dependencies
- future elimination of globals substituted by `require`s
- JS files are watched by webpack which applies a cache. First build is slowish, but subsequent live changes with server running are fast.
- probably lots of other stuff I'm not thinking of

### Additional
- removed `yepnope` from `bower.json`. We deprecated this a long time ago.  `moment` was also removed and is now being `require`d from NPM
